cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/kafka/KafkaOffsetResolver.java

```diff
@@ -0,0 +1,316 @@
+        // the current offset
+        endOffset = offset;
+      }
+      requestTime = eventTime;
+      prevOffset = offset;
+    }
+  }
+
+  /**
+   * Performs a binary search to find the smallest offset of the message between {@code startOffset}
+   * and {@code endOffset} with log event time equal to {@code targetTime}.
+   *
+   * @return the matching offset or {@code -1} if not found
+   */
+  private long searchOffsetByTime(SimpleConsumer consumer, int partition, String clientName, long targetTime,
+                                  long startOffset, long endOffset) {
+    long minOffset = startOffset;
+    long maxOffset = endOffset;
+    long offset;
+
+    while (minOffset <= maxOffset) {
+      offset = minOffset + (maxOffset - minOffset) / 2;
+      long timeStamp = getEventTimeByOffset(consumer, partition, clientName, offset);
+      // if timeStamp is within the range of (targetTime - eventOutOfOrderMillis, targetTime + eventOutOfOrderMillis),
+      // targetTime must be within the range of [timeStamp - eventOutOfOrderMillis, timeStamp + eventOutOfOrderMillis].
+      // Perform a linear search within this range.
+      if (Math.abs(timeStamp - targetTime) < eventOutOfOrderMillis) {
+        return linearSearchAroundOffset(consumer, partition, clientName, offset, targetTime,
+                                        timeStamp - eventOutOfOrderMillis, timeStamp + eventOutOfOrderMillis);
+      }
+      if (timeStamp > targetTime) {
+        maxOffset = offset - 1;
+      }
+      if (timeStamp < targetTime) {
+        minOffset = offset + 1;
+      }
+    }
+    return -1;
+  }
+
+  /**
+   * Performs a linear search to find the smallest offset of the message with log event time
+   * equal to {@code targetTime}. Stop searching when the current message has log event time smaller than
+   * {@code minTime} or larger than {@code maxTime}
+   *
+   * @return the matching offset or {@code -1} if not found
+   */
+  private long linearSearchAroundOffset(SimpleConsumer consumer, int partition, String clientName,
+                                        long startOffset, long targetTime, long minTime, long maxTime) {
+    long smallestMatchingOffset = -1;
+    // search from startOffset and backward first
+    long searchOffset = startOffset;
+    while (searchOffset >= 0) {
+      long time = getEventTimeByOffset(consumer, partition, clientName, searchOffset);
+      if (time < minTime) {
+        break;
+      }
+      if (time == targetTime) {
+        smallestMatchingOffset = searchOffset;
+      }
+      searchOffset--;
+    }
+    if (smallestMatchingOffset != -1) {
+      return smallestMatchingOffset;
+    }
+    searchOffset = startOffset;
+    // The latest offset returned by fetchOffsetByTime is 1 larger than the offset of the latest received message
+    long maxOffset = fetchOffsetByTime(consumer, partition, kafka.api.OffsetRequest.LatestTime());
```

Comment:
Why need to fetch this again? It has been fetched from the `findOffsetByTime` already.